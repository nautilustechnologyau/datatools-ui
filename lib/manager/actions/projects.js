// @flow

import {createAction, type ActionType} from 'redux-actions'
import {browserHistory} from 'react-router'

import {createVoidPayloadAction, secureFetch} from '../../common/actions'
import {getConfigProperty, isModuleEnabled} from '../../common/util/config'
import {deploymentsEnabledAndAccessAllowedForProject} from '../../common/util/permissions'
import {getActiveProject} from '../selectors'
import {
  projectHasAtLeastOneDeployment,
  projectHasAtLeastOneFeedWithAPublishedVersion
} from '../util'
import type {Project} from '../../types'
import type {
  dispatchFn,
  getStateFn,
  FeedSourceTableComparisonColumns,
  FeedSourceTableSortStrategiesWithOrders
} from '../../types/reducers'

import {fetchProjectDeployments} from './deployments'
import {fetchProjectFeeds} from './feeds'
import {handleJobResponse} from './status'
import {setVisibilitySearchText} from './visibilityFilter'

// Bulk Project Actions
const receiveProjects = createAction(
  'RECEIVE_PROJECTS',
  (payload: Array<Project>) => payload
)
const requestingProjects = createVoidPayloadAction('REQUESTING_PROJECTS')
// Single Project Actions
const requestingProject = createVoidPayloadAction('REQUESTING_PROJECT')
const requestingSync = createVoidPayloadAction('REQUESTING_SYNC')
const receiveSync = createVoidPayloadAction('RECEIVE_SYNC')
const savingProject = createVoidPayloadAction('SAVING_PROJECT')
export const setFeedSort = createAction(
  'SET_FEED_SORT_TYPE',
  (payload: FeedSourceTableSortStrategiesWithOrders) => payload
)
// Also used in deployments.js (actions)
export const receiveProject = createAction(
  'RECEIVE_PROJECT',
  (payload: ?Project) => payload
)
const setFeedSourceTableComparisonColumn = createAction(
  'SET_FEED_SOURCE_TABLE_COMPARISON_COLUMN',
  (payload: FeedSourceTableComparisonColumns) => payload
)
export const setFeedSourceTableFilterCountStrategy = createAction(
  'SET_FEED_SOURCE_TABLE_FILTER_COUNT_STRATEGY',
  (payload: string) => payload
)
const runningFetchFeedsForProject = createVoidPayloadAction('RUNNING_FETCH_FEED_FOR_PROJECT')

export type ProjectActions = ActionType<typeof receiveProjects> |
  ActionType<typeof requestingProjects> |
  ActionType<typeof requestingProject> |
  ActionType<typeof requestingSync> |
  ActionType<typeof receiveSync> |
  ActionType<typeof savingProject> |
  ActionType<typeof receiveProject> |
  ActionType<typeof runningFetchFeedsForProject>

export function fetchProjects (getActive: ?boolean = false) {
  return function (dispatch: dispatchFn, getState: getStateFn) {
    dispatch(requestingProjects())
    return dispatch(secureFetch('/api/manager/secure/project'))
      .then(response => response.json())
      .then(projects => {
        dispatch(receiveProjects(projects))
        // return active project if requested
        if (getActive) {
          const activeProject = getActiveProject(getState())
          if (activeProject && !activeProject.feedSources) {
            return dispatch(fetchProjectFeeds(activeProject.id))
              .then(() => activeProject)
          }
        }
        return projects
      })
  }
}

export function fetchProjectsWithPublicFeeds () {
  return function (dispatch: dispatchFn, getState: getStateFn) {
    dispatch(requestingProjects())
    const url = '/api/manager/public/project'
    return dispatch(secureFetch(url))
      .then(response => response.json())
      .then(projects => dispatch(receiveProjects(projects)))
  }
}

export function fetchProject (projectId: string, unsecure: ?boolean) {
  return function (dispatch: dispatchFn, getState: getStateFn) {
    dispatch(requestingProject())
    const apiRoot = unsecure ? 'public' : 'secure'
    const url = `/api/manager/${apiRoot}/project/${projectId}`
    return dispatch(secureFetch(url))
      .then(response => response.json())
      .then(project => {
        dispatch(receiveProject(project))
        return project
      })
  }
}

export function fetchProjectWithFeeds (projectId: string, unsecure: ?boolean, isSaving: ?boolean) {
  return function (dispatch: dispatchFn, getState: getStateFn) {
    dispatch(requestingProject())
    const apiRoot = unsecure ? 'public' : 'secure'
    const url = `/api/manager/${apiRoot}/project/${projectId}`
    return dispatch(secureFetch(url))
      .then(response => response.json())
      .then(project => {
        dispatch(receiveProject(project))
        if (!unsecure) {
          return dispatch(fetchProjectFeeds(project.id))
        }
      })
      .catch(err => {
        console.warn(err)
        dispatch(receiveProject(null))
      })
  }
}

export function deleteProject (project: Project) {
  return function (dispatch: dispatchFn, getState: getStateFn) {
    const url = `/api/manager/secure/project/${project.id}`
    return dispatch(secureFetch(url, 'delete'))
      .then(response => response.json())
      .then(() => browserHistory.push(`/home`))
  }
}

export function updateProject (
  projectId: string,
  changes: {[string]: any},
  fetchFeedsAndDeployments: ?boolean = false
) {
  return function (dispatch: dispatchFn, getState: getStateFn) {
    dispatch(savingProject())
    const url = `/api/manager/secure/project/${projectId}`
    return dispatch(secureFetch(url, 'put', changes))
      .then((res) => {
        if (fetchFeedsAndDeployments) {
          dispatch(fetchProjectWithFeeds(projectId))
          isModuleEnabled('deployment') && dispatch(fetchProjectDeployments(projectId))
        } else {
          return dispatch(fetchProject(projectId))
        }
      })
  }
}

export function deployPublic (project: Project) {
  return function (dispatch: dispatchFn, getState: getStateFn) {
    const url = `/api/manager/secure/project/${project.id}/deployPublic`
    return dispatch(secureFetch(url, 'post'))
      .then((res) => dispatch(handleJobResponse(res, 'Error publishing project feeds')))
  }
}

export function thirdPartySync (projectId: string, type: string) {
  return function (dispatch: dispatchFn, getState: getStateFn) {
    dispatch(requestingSync())
    const url = `/api/manager/secure/project/${projectId}/thirdPartySync/${type}`
    return dispatch(secureFetch(url))
      .then(response => response.json())
      // .catch(err => console.log(err))
      .then(project => {
        dispatch(receiveSync())
        return dispatch(fetchProjectWithFeeds(projectId))
      })
  }
}

export function fetchFeedsForProject (project: Project) {
  return function (dispatch: dispatchFn, getState: getStateFn) {
    dispatch(runningFetchFeedsForProject())
    const url = `/api/manager/secure/project/${project.id}/fetch`
    return dispatch(secureFetch(url, 'post'))
      .then(res => dispatch(handleJobResponse(res, 'Error fetching project feeds')))
  }
}

export function createProject (props: $Shape<Project>) {
  return function (dispatch: dispatchFn, getState: getStateFn) {
    dispatch(savingProject())
    const url = '/api/manager/secure/project'
    let newProjectId
    return dispatch(secureFetch(url, 'post', props))
      .then((res) => res.json())
      .then((json) => {
        newProjectId = json.id
        return dispatch(fetchProjects())
      })
      .then(() => {
        browserHistory.push(`/project/${newProjectId}`)
      })
  }
}

/**
 * Download a merged GTFS file for a Project
 */
export function downloadFeedForProject (project: Project) {
  return function (dispatch: dispatchFn, getState: getStateFn) {
    const url = `/api/manager/secure/project/${project.id}/download`
    return dispatch(secureFetch(url))
      .then(res => dispatch(handleJobResponse(res, 'Error merging project feeds')))
  }
}

/**
 * Download a GTFS file for a merged project feed.
 */
export function downloadMergedFeedViaToken (
  projectId: string,
  isPublic: boolean
) {
  return function (dispatch: dispatchFn, getState: getStateFn) {
    const route = isPublic ? 'public' : 'secure'
    const url = `/api/manager/${route}/project/${projectId}/downloadtoken`
    return dispatch(secureFetch(url))
      .then(response => response.json())
      .then(json => {
        if (getConfigProperty('application.data.use_s3_storage')) {
          // Download object using presigned S3 URL.
          window.location.assign(json.url)
        } else {
          // use token to download feed
          window.location.assign(`/api/manager/downloadprojectfeed/${json.id}`)
        }
      })
  }
}

/**
 * When project viewer component mounts, unconditionally load project with feed
 * sources to ensure that updates to feed sources (e.g., a new feed version) are
 * applied to other project collections (e.g., a deployment that references said
 * feed source).
 */
export async function onProjectViewerMount (projectId: string) {
  return async function (dispatch: dispatchFn, getState: getStateFn) {
    // Remove any feed source search text that may have been previously applied
    dispatch(setVisibilitySearchText(null))
    // Fetch project (and all feeds).
    await dispatch(fetchProjectWithFeeds(projectId))
    const {projects, user} = getState()
    const project = projects.all ? projects.all.find(p => p.id === projectId) : null
    // If deployments are enabled, and the user is a project admin, fetch
    // the deployments right away because this data is needed to calculate
    // whether or not the feed source table comparison column should be
    // shown.
    if (deploymentsEnabledAndAccessAllowedForProject(project, user)) {
      await dispatch(fetchProjectDeployments(projectId))
    }
    await dispatch(calculateFeedSourceTableComparisonColumn(projectId))
  }
}

/**
 * Calculate what comparison column if any should be displayed in the project
 * feed source table
 */
export function calculateFeedSourceTableComparisonColumn (projectId: string) {
  return function (dispatch: dispatchFn, getState: getStateFn) {
    const {projects, user} = getState()
    const {feedSourceTableFilterCountStrategy} = projects.filter
    const project = projects.all
      ? projects.all.find(p => p.id === projectId)
      : null

    if (feedSourceTableFilterCountStrategy === 'PUBLISHED') {
      // this first check is possible because the user must manually set the
      // count strategy after a project is loaded. Therefore, there is
      // gauranteed to be at least one feed source with a published version
      return dispatch(setFeedSourceTableComparisonColumn('PUBLISHED'))
    } else if (
      project &&
        deploymentsEnabledAndAccessAllowedForProject(project, user) &&
        projectHasAtLeastOneDeployment(project)
    ) {
      // deployments are enabled and visible for priveleged user. Show the
      // deployments comparison column if the filter strategy is not set to
      // PUBLISHED
      return dispatch(setFeedSourceTableComparisonColumn('DEPLOYED'))
    } else if (projectHasAtLeastOneFeedWithAPublishedVersion(project)) {
      return dispatch(setFeedSourceTableComparisonColumn('PUBLISHED'))
    } else {
      // deployments not possible, and no published versions, so nothing to
      // compare to
      return dispatch(setFeedSourceTableComparisonColumn(null))
    }
  }
}
