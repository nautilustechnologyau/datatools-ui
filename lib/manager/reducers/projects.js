// @flow

import update from 'react-addons-update'

import {getConfigProperty} from '../../common/util/config'
import {defaultSorter} from '../../common/util/util'
import {getIndexesFromFeed} from '../util'
import type {Action} from '../../types/actions'
import type {ProjectsState} from '../../types/reducers'

export const defaultState = {
  isFetching: false,
  all: [],
  active: null,
  filter: {
    // this controls what comparison column to show
    feedSourceTableComparisonColumn: null,
    // this controls how the feed source table toolbar filter counts are
    // calculated. It can be different than the comparison column in certain
    // cases such as when the filter strategy is for the LATEST feed version
    // but the comparison column is for the DEPLOYED version
    feedSourceTableFilterCountStrategy: 'LATEST',
    // in the feed source table, this is a validation status filter
    filter: null,
    // This is used to sort both the Project List or the FeedSourceTable
    searchText: null,
    // Used by the label filtering dropdown
    labelsFilterMode: 'any',
    labels: []
  },
  // this is used to sort the FeedSourceTable
  sort: 'alphabetically-asc'
}

/* eslint-disable complexity */
const projects = (state: ProjectsState = defaultState, action: Action): ProjectsState => {
  switch (action.type) {
    case 'SET_FEED_SORT_TYPE':
      return update(state, {sort: {$set: action.payload}})
    case 'SET_PROJECT_VISIBILITY_SEARCH_TEXT':
      return update(state, {filter: {searchText: {$set: action.payload}}})
    case 'SET_PROJECT_VISIBILITY_LABEL':
      return update(state, {filter: {labels: {$set: action.payload}}})
    case 'SET_PROJECT_VISIBILITY_LABEL_MODE':
      return update(state, {filter: {labelsFilterMode: {$set: action.payload}}})
    case 'SET_PROJECT_VISIBILITY_FILTER':
      return update(state, {filter: {filter: {$set: action.payload}}})
    case 'CREATE_DEPLOYMENT':
      const projectIndex = state.all.findIndex(p => p.id === action.payload)
      const project = state.all[projectIndex]
      let newState
      // if project's deployment array is undefined, add it
      if (!project.deployments) {
        newState = update(state, {
          all: {[projectIndex]: {$merge: {deployments: []}}}
        })
      }
      // add new empty feed source to feedSources array
      const deployment = {
        isCreating: true,
        name: '',
        project: project
      }
      return update(newState || state, {
        all: {[projectIndex]: {deployments: {$unshift: [deployment]}}}
      })
    case 'REQUESTING_FEEDSOURCE':
    case 'REQUESTING_FEEDSOURCES':
    case 'REQUESTING_PROJECT':
    case 'REQUEST_PROJECTS': {
      return update(state, { isFetching: { $set: true } })
    }
    case 'RECEIVE_PROJECTS': {
      const activeProjectId = state.active
        ? state.active.id
        : getConfigProperty('application.active_project')
      let activeIndex = action.payload.findIndex(p => p.id === activeProjectId)
      if (activeIndex === -1) {
        activeIndex = 0
      }
      return {
        isFetching: false,
        all: action.payload,
        active: action.payload[activeIndex],
        filter: {
          feedSourceTableComparisonColumn: null,
          feedSourceTableFilterCountStrategy: 'LATEST',
          filter: null,
          labels: [],
          searchText: null,
          labelsFilterMode: 'any'
        },
        sort: 'alphabetically-asc'
      }
    }
    case 'RECEIVE_PROJECT': {
      let projects, projectIndex
      if (!action.payload) {
        return update(state, { isFetching: { $set: false } })
      }
      if (!state.all) {
        // there are no current projects loaded
        projects = [action.payload]
      } else {
        // projects already loaded
        // extract because flow https://flow.org/en/docs/lang/refinements/#toc-refinement-invalidations
        const {id} = action.payload
        projectIndex = state.all.findIndex(p => p.id === id)
        if (projectIndex === -1) {
          // projects loaded but not this one; add it
          projects = [
            ...state.all,
            action.payload
          ]
        } else {
          // FIXME: Use $splice
          // projects loaded including this one, replace it
          projects = [
            ...state.all.slice(0, projectIndex),
            action.payload,
            ...state.all.slice(projectIndex + 1)
          ]
        }
      }
      const activeProjectId = state.active
        ? state.active.id
        : getConfigProperty('application.active_project')
      let activeProject = action.payload.id === activeProjectId
        ? action.payload
        : null
      if (state.active && !activeProject) {
        // Active project already exists and received project does not match
        // active project.
        activeProject = state.active
      }
      return update(state, {
        isFetching: {$set: false},
        active: {$set: activeProject},
        all: {$set: projects}
      })
    }
    case 'RECEIVE_FEEDSOURCES': {
      // extract because flow https://flow.org/en/docs/lang/refinements/#toc-refinement-invalidations
      const {feedSources, projectId} = action.payload
      const {projectIndex} = getIndexesFromFeed({ projectId, state })
      if (projectIndex < 0) {
        console.warn(`Project ID ${projectId} not found in state`)
        return state
      }
      const currentFeedSources = state.all[projectIndex].feedSources
      let sourceIndex, feeds
      if (currentFeedSources && currentFeedSources.length === 1) {
        // If lazy fetching the rest of the feed sources, don't overwrite current
        // feed source.
        sourceIndex = feedSources
          .findIndex(fs => fs.id === currentFeedSources[0].id)
        if (sourceIndex === -1) {
          // There is no feed source to overwrite (none is loaded or none exist).
          feeds = feedSources
        } else {
          feedSources.splice(sourceIndex, 1)
          feeds = [
            ...currentFeedSources,
            ...feedSources
          ]
        }
      } else {
        feeds = action.payload.feedSources
      }
      feeds = feeds.sort(defaultSorter)
      if (state.active && action.payload.projectId === state.active.id) {
        // Project is active
        return update(state, {
          isFetching: {$set: false},
          active: {feedSources: {$set: feeds}},
          all: {
            [projectIndex]: {feedSources: {$set: feeds}}
          }
        })
      } else {
        // If projectId does not match active project
        return update(state, {
          isFetching: {$set: false},
          all: {
            [projectIndex]: {feedSources: {$set: feeds}}
          }
        })
      }
    }
    case 'RECEIVE_FEEDSOURCE': {
      if (!action.payload) {
        return update(state, {isFetching: {$set: false}})
      }
      // extract because flow https://flow.org/en/docs/lang/refinements/#toc-refinement-invalidations
      const {id, projectId} = action.payload
      const {projectIndex, sourceIndex} = getIndexesFromFeed({
        feedSourceId: id,
        projectId,
        state
      })
      const updatedSources = sourceIndex > -1
        ? {[sourceIndex]: {$set: action.payload}}
        : {$set: [action.payload]}
      return update(state, {
        all: {
          [projectIndex]: {
            feedSources: updatedSources // contains $set update
          }
        },
        isFetching: { $set: false }
      })
    }
    case 'RECEIVE_FEEDVERSIONS': {
      const {feedSource, versions} = action.payload
      const {projectIndex, sourceIndex} = getIndexesFromFeed({
        feedSourceId: feedSource.id,
        projectId: feedSource.projectId,
        state
      })
      const versionSort = (a, b) => {
        if (a.version < b.version) return -1
        if (a.version > b.version) return 1
        return 0
      }
      return update(state, {
        all: {[projectIndex]: {feedSources: {[sourceIndex]: {$merge: {
          feedVersions: versions.map(() => null),
          feedVersionSummaries: versions.sort(versionSort)
        }}}}}
      })
    }
    case 'RECEIVE_FEEDVERSION': {
      const {feedSource, id} = action.payload
      const {projectIndex, sourceIndex, versionIndex} = getIndexesFromFeed({
        feedSourceId: feedSource.id,
        feedVersionId: id,
        projectId: feedSource.projectId,
        state
      })
      const updatedVersions = versionIndex > -1
        ? {[versionIndex]: {$set: action.payload}}
        : {$set: [action.payload]}
      return update(state, {
        all: {[projectIndex]: {feedSources: {[sourceIndex]: {
          feedVersions: updatedVersions
        }}}}
      })
    }
    case 'PUBLISHED_FEEDVERSION': {
      const {feedSource, id} = action.payload
      const {projectIndex, sourceIndex, versionIndex} = getIndexesFromFeed({
        feedSourceId: feedSource.id,
        feedVersionId: id,
        projectId: feedSource.projectId,
        state
      })
      if (versionIndex === -2) {
        console.warn('Feedsource has no feed version')
        return state
      }
      return update(state, {
        all: {[projectIndex]: {feedSources: {[sourceIndex]: {feedVersions: {
          [versionIndex]: {$set: action.payload
          }}}}}}
      })
    }
    case 'RECEIVE_VALIDATION_ISSUE_COUNT': {
      const {feedVersion} = action.payload
      const {feedSource, id} = feedVersion
      const {projectIndex, sourceIndex, versionIndex} = getIndexesFromFeed({
        feedSourceId: feedSource.id,
        feedVersionId: id,
        projectId: feedSource.projectId,
        state
      })
      if (versionIndex === -2) {
        console.warn('Feedsource has no feed version')
        return state
      }
      if (!action.payload || !action.payload.validationIssueCount) return state
      return update(state, {
        all: {[projectIndex]: {feedSources: {[sourceIndex]: {feedVersions: {
          [versionIndex]: {validationResult: {$merge: action.payload.validationIssueCount}
          }}}}}}
      })
    }
    case 'RECEIVE_VALIDATION_ERRORS': {
      const {errorType, feedVersion} = action.payload
      const {feedSource} = feedVersion
      const {errorIndex, projectIndex, sourceIndex, versionIndex} = getIndexesFromFeed({
        errorType,
        feedSourceId: feedSource.id,
        feedVersionId: feedVersion.id,
        projectId: feedSource.projectId,
        state
      })
      if (
        versionIndex === -2 ||
        errorIndex === -2 ||
        !state.all[projectIndex].feedSources ||
        !state.all[projectIndex].feedSources[sourceIndex].feedVersions
      ) {
        console.warn('Feedsource has no feed version')
        return state
      }
      const {validationResult} = state.all[projectIndex].feedSources[sourceIndex].feedVersions[versionIndex]
      if (!validationResult.error_counts) {
        console.warn('validation result does not contain error_counts array!')
        return state
      }
      const newErrors = validationResult.error_counts[errorIndex].errors
        ? {$push: action.payload.errors}
        : {$set: action.payload.errors}
      return update(state, {
        all: {[projectIndex]: {feedSources: {[sourceIndex]: {feedVersions: {
          [versionIndex]: {validationResult: {error_counts: {
            [errorIndex]: {errors: newErrors}
          }}}}}}}
        }
      })
    }
    case 'RECEIVE_FEEDVERSION_ISOCHRONES': {
      const {feedSource, feedVersion} = action.payload
      const {projectIndex, sourceIndex, versionIndex} = getIndexesFromFeed({
        feedSourceId: feedSource.id,
        feedVersionId: feedVersion.id,
        projectId: feedSource.projectId,
        state
      })
      if (versionIndex === -2) {
        console.warn('Feedsource has no feed version')
        return state
      }
      const {date, fromLat, fromLon, fromTime, isochrones, toTime} = action.payload
      isochrones.properties = { fromLat, fromLon, date, fromTime, toTime }
      isochrones.features = isochrones && isochrones.features
        ? isochrones.features.map(f => {
          f.type = 'Feature'
          return f
        })
        : null
      return update(state, {
        all: {[projectIndex]: {feedSources: {[sourceIndex]: {feedVersions: {
          [versionIndex]: {$merge: {isochrones: isochrones
          }}}}}}}
      })
    }
    case 'RECEIVE_DEPLOYMENTS': {
      const {deployments: deploymentsSummaries, feedSourceId, projectId} = action.payload
      // Deployment summaries don't include a project id, so we need to add one
      // Deployment summaries don't include feedversions array, so we create a fake array that gets replaced later
      const deployments = deploymentsSummaries.map(d => ({...d, projectId, feedVersions: Array(d.numberOfFeedVersions)}))

      const {projectIndex} = getIndexesFromFeed({ projectId, state })
      if (feedSourceId && state.all[projectIndex].feedSources) {
        const feedSourceIndex = state.all[projectIndex].feedSources
          .findIndex(fs => fs.id === feedSourceId)
        return update(state, {
          all: {[projectIndex]: {
            feedSources: {[feedSourceIndex]: {$merge: {deployments}}}
          }}
        })
      }
      if (state.active && projectId === state.active.id) {
        return update(state, {
          active: {$merge: {deployments}},
          all: {[projectIndex]: {$merge: {deployments}}}
        })
      } else {
        // if projectId does not match active project
        return update(state, {
          all: {[projectIndex]: {$merge: {deployments}}}
        })
      }
    }
    case 'RECEIVE_DEPLOYMENT': {
      const {projectIndex} = getIndexesFromFeed({
        projectId: action.payload.projectId,
        state
      })
      const existingDeployments = state.all[projectIndex].deployments || []
      let updatedDeployments
      const deploymentIndex = existingDeployments
        .findIndex(s => s.id === action.payload.id)
      if (deploymentIndex === -1) {
        // Deployment does not currently exist; add it.
        updatedDeployments = [
          ...existingDeployments,
          action.payload
        ]
      } else {
        // Existing deployment array includes this one, replace it
        updatedDeployments = [
          ...existingDeployments.slice(0, deploymentIndex),
          action.payload,
          ...existingDeployments.slice(deploymentIndex + 1)
        ]
      }
      return update(state, {
        all: {[projectIndex]: {$merge: {deployments: updatedDeployments}}}
      })
    }
    case 'RECEIVE_NOTES_FOR_FEEDSOURCE': {
      const {feedSource, notes} = action.payload
      const {projectIndex, sourceIndex} = getIndexesFromFeed({
        feedSourceId: feedSource.id,
        projectId: feedSource.projectId,
        state
      })
      return update(state, {
        // Merge notes into feed source and ensure note count matches notes length
        all: {[projectIndex]: {feedSources: {
          [sourceIndex]: {$merge: {notes, noteCount: notes.length}}}}}
      })
    }
    case 'RECEIVE_NOTES_FOR_FEEDVERSION': {
      const {feedVersion, notes} = action.payload
      const {feedSource} = feedVersion
      const {projectIndex, sourceIndex, versionIndex} = getIndexesFromFeed({
        feedSourceId: feedSource.id,
        feedVersionId: feedVersion.id,
        projectId: feedSource.projectId,
        state
      })
      if (versionIndex === -2) {
        console.warn('Feedsource has no feed version')
        return state
      }
      return update(state, {all: {
        [projectIndex]: {feedSources: {[sourceIndex]: {feedVersions: {
          // Merge notes into feed version and ensure note count matches notes length
          [versionIndex]: {$merge: {notes, noteCount: notes.length}
          }}}}}}
      })
    }
    case 'RECEIVE_GTFSEDITOR_SNAPSHOTS': {
      const {feedSource} = action.payload
      const {projectIndex, sourceIndex} = getIndexesFromFeed({
        feedSourceId: feedSource.id,
        projectId: feedSource.projectId,
        state
      })
      return update(state, {
        all: {
          [projectIndex]: {
            feedSources: {
              [sourceIndex]: {
                $merge: {
                  editorSnapshots: action.payload.snapshots
                }
              }
            }
          }
        }
      })
    }
    case 'SET_FEED_SOURCE_TABLE_FILTER_COUNT_STRATEGY':
      return update(
        state,
        {
          filter: {
            feedSourceTableFilterCountStrategy: {
              $set: action.payload
            }
          }
        }
      )
    case 'SET_FEED_SOURCE_TABLE_COMPARISON_COLUMN':
      return update(
        state,
        {
          filter: {
            feedSourceTableComparisonColumn: {
              $set: action.payload
            }
          }
        }
      )
    default:
      return state
  }
}

export default projects
