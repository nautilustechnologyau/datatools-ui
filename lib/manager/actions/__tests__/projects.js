// @flow

import nock from 'nock'

import mockData from '../../../../__tests__/test-utils/mock-data'
import * as projectsActions from '../projects'

const {
  getMockInitialState,
  getMockStateWithAdminUser,
  makeMockStore
} = mockData.store

describe('manager > actions > projects >', () => {
  it('should set a new feed source sort order', () => {
    const store = makeMockStore(getMockInitialState())
    store.dispatch(projectsActions.setFeedSort('alphabetically-desc'))
    store.expectStateToMatchSnapshot('projects')
  })

  it('should load a project and associated data', async () => {
    const projectId = 'mock-project-with-deployments-id'
    const serverUrl = 'http://localhost:4000'
    // mock for fetching project
    nock(serverUrl)
      .get(`/api/manager/secure/project/${projectId}`)
      .reply(200, mockData.manager.mockProjectWithDeploymentUnloaded)
      // mock for fetching project feeds
      .get(`/api/manager/secure/feedsourceSummaries?projectId=${projectId}`)
      .reply(200, [mockData.manager.mockFeedSourceSummaryWithVersion])
      // mock for fetching the project deployments
      .get(`/api/manager/secure/deployments?projectId=${projectId}`)
      .reply(200, [mockData.manager.mockDeployment])
      .get(`/api/manager/secure/deploymentSummaries?projectId=${projectId}`)
      .reply(200, [mockData.manager.mockDeploymentSummary])

    const store = makeMockStore(getMockStateWithAdminUser())
    await store.dispatch(
      projectsActions.onProjectViewerMount(projectId)
    )
    store.expectStateToMatchSnapshot('projects')
  })
})
