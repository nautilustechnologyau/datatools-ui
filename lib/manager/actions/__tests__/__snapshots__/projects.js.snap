// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manager > actions > projects > should load a project and associated data 1`] = `
Object {
  "active": null,
  "all": Array [
    Object {
      "autoDeploy": false,
      "autoDeployTypes": Array [],
      "autoDeployWithCriticalErrors": false,
      "autoFetchFeeds": true,
      "autoFetchHour": 0,
      "autoFetchMinute": 0,
      "bounds": null,
      "buildConfig": Object {
        "fares": null,
        "fetchElevationUS": null,
        "stationTransfers": null,
        "subwayAccessTime": null,
      },
      "dateCreated": 1526305321824,
      "defaultTimeZone": null,
      "deployments": Array [
        Object {
          "dateCreated": 1553292345720,
          "deployedTo": null,
          "feedVersions": Array [
            undefined,
          ],
          "id": "mock-deployment-id-1",
          "lastDeployed": null,
          "name": "mock-deployment",
          "projectId": "mock-project-with-deployments-id",
        },
      ],
      "feedSources": Array [
        Object {
          "deployable": false,
          "id": "mock-feed-with-version-id",
          "isPublic": false,
          "labelIds": Array [],
          "lastUpdated": 1543389038810,
          "latestValidation": Object {
            "agencies": null,
            "agencyCount": 1,
            "avgDailyRevenueTime": 0,
            "bounds": Object {
              "east": -76.481211,
              "north": 39.0486949672717,
              "south": 38.92884,
              "west": -76.5673055566884,
            },
            "endDate": "20190801",
            "errorCount": 78,
            "feedVersionId": "mock-feed-version-id",
            "loadFailureReason": null,
            "loadStatus": "SUCCESS",
            "routeCount": 10,
            "startDate": "20180801",
            "stopCount": 237,
            "stopTimesCount": 11170,
            "tripCount": 415,
          },
          "name": "test feed with a version",
          "projectId": "mock-project-id",
        },
      ],
      "id": "mock-project-with-deployments-id",
      "labels": Array [],
      "lastUpdated": 1553236399556,
      "name": "mock-project-with-deployments",
      "organizationId": null,
      "otpServers": Array [],
      "peliasWebhookUrl": null,
      "pinnedDeploymentId": null,
      "routerConfig": Object {
        "driveDistanceReluctance": null,
        "itineraryFilters": Object {
          "nonTransitGeneralizedCostLimit": null,
        },
        "requestLogFile": null,
        "stairsReluctance": null,
        "updaters": null,
        "walkSpeed": null,
      },
      "sharedStopsConfig": null,
      "useCustomOsmBounds": false,
      "user": null,
    },
  ],
  "filter": Object {
    "feedSourceTableComparisonColumn": "DEPLOYED",
    "feedSourceTableFilterCountStrategy": "LATEST",
    "filter": null,
    "labels": Array [],
    "labelsFilterMode": "any",
    "searchText": null,
  },
  "isFetching": false,
  "sort": "alphabetically-asc",
}
`;

exports[`manager > actions > projects > should set a new feed source sort order 1`] = `
Object {
  "active": null,
  "all": Array [],
  "filter": Object {
    "feedSourceTableComparisonColumn": null,
    "feedSourceTableFilterCountStrategy": "LATEST",
    "filter": null,
    "labels": Array [],
    "labelsFilterMode": "any",
    "searchText": null,
  },
  "isFetching": false,
  "sort": "alphabetically-desc",
}
`;
